res/controllers/Traktor-Kontrol-S2-MK1-hid-scripts.js

```diff
@@ -15,1349 +15,1352 @@
-    TraktorS2MK1.previousBrowse = field.value;
+    shutdown() {
+        const data = [];
+        for (let i = 0; i < 61; i++) {
+            data[i] = 0;
+        }
+        // light up warning light
+        data[0x33 - 1] = ButtonBrightnessOn;
+        controller.send(data, data.length, 0x80);
+    }
 
-    if (TraktorS2MK1.shiftPressed["[Channel1]"] || TraktorS2MK1.shiftPressed["[Channel2]"]) {
-        delta *= 5;
+    incomingData(data, length) {
+        this.controller.parsePacket(data, length);
     }
-    engine.setValue("[Playlist]", "SelectTrackKnob", delta);
-};
+    // The input report 0x01 handles buttons and jog wheels.
+    inputReport0x01Callback(packet, data) {
+        for (const name in data) {
+            const field = data[name];
+            if (field.name === "!jog_wheel") {
+                this.controller.processControl(field);
+                continue;
+            }
 
-TraktorS2MK1.scalerParameter = function(group, name, value) {
-    return script.absoluteLin(value, 0, 1, 16, 4080);
-};
+            this.controller.processButton(field);
+        }
+    }
+    // There are no buttons handled by input report 0x02, so this is a little simpler.
+    inputReport0x02Callback(packet, data) {
+        for (const name in data) {
+            const field = data[name];
+            this.controller.processControl(field);
+        }
+    }
+    samplerGainKnob(field) {
+        for (let i = 1; i <= 8; i++) {
+            engine.setParameter("[Sampler" + i + "]", "pregain", field.value / 4095);
+        }
+    }
 
-TraktorS2MK1.scalerVolume = function(group, name, value) {
-    if (group === "[Master]") {
-        return script.absoluteNonLin(value, 0, 1, 4, 16, 4080);
-    } else {
-        return script.absoluteNonLin(value, 0, 0.25, 1, 16, 4080);
+    toggleButton(field) {
+        if (field.value > 0) {
+            script.toggleControl(field.group, field.name);
+        }
     }
-};
 
-TraktorS2MK1.scalerSlider = function(group, name, value) {
-    return script.absoluteLin(value, -1, 1, 16, 4080);
-};
+    browseEncoder(field) {
+        const delta = encoderDirection(field.value, this.previousBrowse);
+        this.previousBrowse = field.value;
 
-TraktorS2MK1.outputChannelCallback = function(value, group, key) {
-    var ledValue = 0x05;
-    if (value) {
-        ledValue = 0x1F;
+        engine.setValue("[Library]", "MoveVertical", delta);
     }
-    TraktorS2MK1.controller.setOutput(group, key, ledValue, !TraktorS2MK1.batchingLEDUpdate);
-};
 
-TraktorS2MK1.outputChannelCallbackDark = function(value, group, key) {
-    var ledValue = 0x00;
-    if (value) {
-        ledValue = 0x1F;
+    browseEncoderPress(field) {
+        if (this.shiftPressed()) {
+            engine.setValue("[Library]", "GoToItem", field.value);
+        } else {
+            if (field.value > 0) {
+                script.toggleControl("[Master]", "maximize_library");
+            }
```

Comment:
The other way around makes more sense, no?
`[Library], GoToItem` can be used in many places, so I'd say it's used much more often than maximizing the library.

Btw there is no control to move the library focus to the search box? In Mixxx 2.4 the search box is now a combobox with saved search queries, and using that with controllers is nice new feature.

Reply:
When holding shift the focus is on the search box (I assume you mean the box with Tracks, Auto-DJ, …). When shift is not pressed (focus on Track list), GoToItem just loads the track, so it can be something different like maximize library. (this is also the default behavior in Traktor).